crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -1358,61 +1418,48 @@ impl<TX: DbTx> BlockReader for DatabaseProvider<TX> {
+        })
+    }
 
-        for num in range {
-            if let Some((_, header)) = headers_cursor.seek_exact(num)? {
-                // If the body indices are not found, this means that the transactions either do
-                // not exist in the database yet, or they do exit but are
-                // not indexed. If they exist but are not indexed, we don't
-                // have enough information to return the block anyways, so
-                // we skip the block.
-                if let Some((_, block_body_indices)) = block_body_cursor.seek_exact(num)? {
-                    let tx_range = block_body_indices.tx_num_range();
-                    let body = if tx_range.is_empty() {
-                        Vec::new()
-                    } else {
-                        self.transactions_by_tx_range_with_cursor(tx_range, &mut tx_cursor)?
-                            .into_iter()
-                            .map(Into::into)
-                            .collect()
-                    };
+    fn block_with_senders_range(
+        &self,
+        range: RangeInclusive<BlockNumber>,
+    ) -> ProviderResult<Vec<BlockWithSenders>> {
+        let mut tx_cursor = self.tx.cursor_read::<tables::Transactions>()?;
+        let mut senders_cursor = self.tx.cursor_read::<tables::TransactionSenders>()?;
 
-                    // If we are past shanghai, then all blocks should have a withdrawal list,
-                    // even if empty
-                    let withdrawals =
-                        if self.chain_spec.is_shanghai_active_at_timestamp(header.timestamp) {
-                            Some(
-                                withdrawals_cursor
-                                    .seek_exact(num)?
-                                    .map(|(_, w)| w.withdrawals)
-                                    .unwrap_or_default(),
-                            )
-                        } else {
-                            None
-                        };
-                    let ommers = if self.chain_spec.final_paris_total_difficulty(num).is_some() {
-                        Vec::new()
-                    } else {
-                        ommers_cursor.seek_exact(num)?.map(|(_, o)| o.ommers).unwrap_or_default()
-                    };
+        self.process_block_range(range, |tx_range, header, ommers, withdrawals| {
+            let (body, senders) = if tx_range.is_empty() {
+                (Vec::new(), Vec::new())
+            } else {
+                let body = self
+                    .transactions_by_tx_range_with_cursor(tx_range.clone(), &mut tx_cursor)?
+                    .into_iter()
+                    .map(Into::into)
+                    .collect();
+                // fetch senders from the senders table
+                let senders = senders_cursor
+                    .walk_range(tx_range)?
+                    .map(|entry| entry.map(|(_, sender)| sender))
+                    .collect::<Result<Vec<_>, _>>()?;
```

Comment:
senders might be missing from the table if they are pruned, there other provider functions that can be used to get them, so you might need to do a mix of fetching what you can, and recover the rest (take a look at `senders_by_tx_range`)

Reply:
otherwise, if they are pruned `try_with_senders_unchecked` will  be recalculating every send from the body.

If the above comment suggestion seems too complex, you could just let `try_with_senders_unchecked` calculate the senders by passing an empty vec. It can be improved on another PR

Reply:
Never mind, will give it a try later.

Reply:
Done.